### PR TITLE
Implement Scaling Provider Factory Method Pattern

### LINKED ISSUE #180: Implement Cluster Scaling Multi-Provider Model
**Description**

Replicator needs to move to a multi-provider model to implement support for various cluster scaling backends (e.g. AWS, GCE, Terraform, Generic, etc). 

This feature request will cover the work needed to restructure the cluster scaling backends into a multi-provider factory pattern.

### PR DESCRIPTION
**Pull request by ericwestfall**
This commit introduces a factory method pattern and a new package, 
`provider`, that implements all cloud provider specific functionality for 
scaling operations. This model will pave the way for supporting multiple 
cloud providers and possibly other more generic scaling backends.

This change moves the worker pool state tracking object into the
worker pool object and allows us to track the desired scaling
direction to simplify the number of objects that need to be
passed around between the core and the providers.

We also introduce a new state field that allows the core to register
nodes as eligible for targeted scaling operations. At present, this
is used to specify nodes that have been identified as low utilization,
have been placed in drain mode and are eligible for detachment and
termination.

The core cluster scaling methods have been updated to use very
fine-grained locking and no longer block the node discovery thread
from updating the node registry while cluster scaling operations are
underway. This allows us to monitor the node registry to confirm a
newly launched instance has successfully joined the cluster while
the scaling operation is still in progress instead of explicitly calling 
Nomad API endpoints.

Finally, this commit introduces tracking of node registration
timestamp information to allow methods to understand in what order
nodes within a worker pool were registered.

Closes #180

### CODE REVIEW

File: provider/provider.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	"strings"
+
+	"github.com/elsevier-core-engineering/replicator/logging"
+	"github.com/elsevier-core-engineering/replicator/replicator/structs"
+)
+
+// BuiltinScalingProviders tracks the available scaling providers.
+// The provider name is the name used when configuring nodes for autoscaling.
+var BuiltinScalingProviders = map[string]ScalingProviderFactory{
+	"aws": NewAwsScalingProvider,
```
[the remark this answers is not in the document]
Comment (jippi) [reply provider/provider.go]: yes, it should be one of the two ;)

[the remark this answers is not in the document]
Comment (ericwestfall) [reply provider/provider.go]: Something about `AWSScalingProvider` just hits the eyes wrong imo .. I'm open to changing it but the neurotic part of me won't like it

**Comment by ericwestfall** [comment]
Alright @jrasell and @jippi, I've updated the package to be called `cloudprovider` and while I was at it I decided to give a little more structure to the `aws` cloud provider by adding it to its own package and splitting the methods up a bit. 

All the commented out imports have been removed. 

@jrasell I'm open to renamed `AwsScalingProvider` but I just hate the way `AWSScalingProvider` looks. I'm open to whatevs just my $0.02.

**Comment by ericwestfall** [comment]
Closing in favor of #203, where the massive conflicts have been dealt with.

**Pull request closed without merging**